hrm-service/migrations/20220415170355-create-Audits.js

```diff
@@ -0,0 +1,143 @@
+'use strict';
+
+module.exports = {
+  up: async queryInterface => {
+    await queryInterface.sequelize.query(
+      `CREATE SEQUENCE IF NOT EXISTS public."Audits_id_seq"
+        INCREMENT 1
+        START 1
+        MINVALUE 1
+        MAXVALUE 9223372036854775807
+        CACHE 1;`,
+    );
+    console.log('Sequence "Audits_id_seq" created');
+
+    await queryInterface.sequelize.query(`ALTER SEQUENCE public."Audits_id_seq" OWNER TO hrm`);
+    console.log('Sequence "Audits_id_seq" ownership altered.');
+
+    await queryInterface.sequelize.query(
+      `CREATE TABLE IF NOT EXISTS public."Audits"
+      (
+        id integer NOT NULL DEFAULT nextval('"Audits_id_seq"'::regclass),
+        "tableName" text NOT NULL,
+        "targetPK" text NOT NULL,
+        operation text NOT NULL,
+        "updatedBy" text,
+        "previousRecord" jsonb,
+        timestamp_trx timestamp with time zone NOT NULL,
+        timestamp_stm timestamp with time zone NOT NULL,
+        timestamp_clock timestamp with time zone NOT NULL,
+        PRIMARY KEY (id)
+      );`,
+    );
+    console.log('Table "Audits" created');
+
+    await queryInterface.sequelize.query(`ALTER TABLE IF EXISTS public."Audits" OWNER TO hrm`);
+    console.log('Table "Audits" ownership altered.');
+
+    await queryInterface.sequelize.query(`
+      CREATE FUNCTION public.audit_trigger()
+        RETURNS trigger
+        LANGUAGE 'plpgsql'
+        NOT LEAKPROOF
+      AS $BODY$
+      DECLARE
+        audit_row public."Audits";
+        "targetPK" text;
+        "updatedBy" text;
+      BEGIN
+        IF TG_WHEN <> 'AFTER' THEN
+          RAISE EXCEPTION 'audit_trigger() may only run as an AFTER trigger';
+        END IF;
+
+        IF (TG_OP = 'UPDATE' AND TG_LEVEL = 'ROW') THEN
```

Comment:
In this case I'm not considering a potential change on the PK. Is that something we should look for? We might lose the link between a record and it's audits if one of the keys changes.

Reply:
I don't think this is an issue, as long as the change itself is recorded.

I do just wonder if it's worth recording the PK and updatedBy fields separately - it seems to increase the complexity & brittleness of the trigger logic somewhat. Would it not be simpler just to record the whole NEW row as a jsonb blob alongside the old one? Then we're covered for any shape of table we care to audit without updating the logic?